cli/firmware/flash.go

```diff
@@ -159,82 +161,96 @@ func run(cmd *cobra.Command, args []string) {
+			feedback.Errorf("Error during firmware flashing: %s", err)
 			os.Exit(errorcodes.ErrGeneric)
 		}
-		if newUploadPort != "" {
-			logrus.Infof("Found port to upload Loader: %s", newUploadPort)
-			uploadPort = newUploadPort
+		if err != nil {
+			feedback.Errorf("Error during firmware flashing: %s", err)
+			continue
 		}
-	}
 
-	// Flash loader Sketch
-	programmerOut := new(bytes.Buffer)
-	programmerErr := new(bytes.Buffer)
-	if feedback.GetFormat() == feedback.JSON {
-		err = programmer.Flash(commandLine, programmerOut, programmerErr)
-	} else {
-		err = programmer.Flash(commandLine, os.Stdout, os.Stderr)
-	}
-	if err != nil {
-		feedback.Errorf("Error during firmware flashing: %s", err)
-		os.Exit(errorcodes.ErrGeneric)
-	}
-
-	// Wait a bit after flashing the loader sketch for the board to become
-	// available again.
-	time.Sleep(2 * time.Second)
-
-	// Get flasher depending on which module to use
-	var f flasher.Flasher
-	switch moduleName {
-	case "NINA":
-		f, err = flasher.NewNinaFlasher(uploadPort)
-	case "SARA":
-		f, err = flasher.NewSaraFlasher(uploadPort)
-	case "WINC1500":
-		f, err = flasher.NewWincFlasher(uploadPort)
-	default:
-		err = fmt.Errorf("unknown module: %s", moduleName)
-	}
-	if err != nil {
-		feedback.Errorf("Error during firmware flashing: %s", err)
-		os.Exit(errorcodes.ErrGeneric)
-	}
-	defer f.Close()
+		// now flash the actual firmware
+		flasherOut := new(bytes.Buffer)
+		flasherErr := new(bytes.Buffer)
+		if feedback.GetFormat() == feedback.JSON {
+			err = f.FlashFirmware(firmwareFile, flasherOut)
+		} else {
+			err = f.FlashFirmware(firmwareFile, os.Stdout)
+		}
+		if err != nil {
+			feedback.Errorf("Error during firmware flashing: %s", err)
+			flasherErr.Write([]byte(fmt.Sprintf("Error during firmware flashing: %s", err)))
+		}
+		f.Close()
 
-	// now flash the actual firmware
-	flasherOut := new(bytes.Buffer)
-	flasherErr := new(bytes.Buffer)
-	if feedback.GetFormat() == feedback.JSON {
-		err = f.FlashFirmware(firmwareFile, flasherOut)
-	} else {
-		err = f.FlashFirmware(firmwareFile, os.Stdout)
-	}
-	if err != nil {
-		feedback.Errorf("Error during firmware flashing: %s", err)
-		flasherErr.Write([]byte(fmt.Sprintf("Error during firmware flashing: %s", err)))
-	}
+		// Print the results
+		feedback.PrintResult(&flasher.FlashResult{
+			Programmer: (&flasher.ExecOutput{
+				Stdout: programmerOut.String(),
+				Stderr: programmerErr.String(),
+			}),
+			Flasher: (&flasher.ExecOutput{
+				Stdout: flasherOut.String(),
+				Stderr: flasherErr.String(),
+			}),
+		})
 
-	// Print the results
-	feedback.PrintResult(&flasher.FlashResult{
-		Programmer: (&flasher.ExecOutput{
-			Stdout: programmerOut.String(),
-			Stderr: programmerErr.String(),
-		}),
-		Flasher: (&flasher.ExecOutput{
-			Stdout: flasherOut.String(),
-			Stderr: flasherErr.String(),
-		}),
-	})
-	// Exit if something went wrong but after printing
-	if err != nil {
-		os.Exit(errorcodes.ErrGeneric)
+		if err == nil {
```

Comment:
Is this here only as a "belt and suspenders" measure in case there was some missed error handling in the code?

Reply:
Actually, this condition is here because it allows breaking from the retry loop. Not sure if this can be handled better...
After every error in the loop there's a continue statement that allows to start from the next iteration of the loop in case something is wrong

Reply:
>After every error in the loop there's a continue statement that allows to start from the next iteration of the loop in case something is wrong

And for that reason, `err` should never have any value other than `nil` if this condition is reached, right?

Reply:
Correct

Reply:
:see_no_evil: I forgot the continue statement here https://github.com/arduino/FirmwareUploader/pull/61/files#diff-9bcc47c1909291c8177f70e16def7a94afe0d6c3fb6c937a8f2ccc91432b8aecR233
Thanks a lot @per1234

Reply:
OK, I'm mostly just trying to make sure I understand the code. Since it should never be other than `nil` if the code is working correctly, I was wondering if a better "belt and suspenders" approach wouldn't be to panic:
```go
if err !=nil{
  panic(fmt.Sprintf("Unhandled error: %s", err))
}
break
```
I generally just throw in a panic to handle situations that should never occur unless something is broken in the code. It's not clear to me whether it would be beneficial to make the code resilient to this condition, since it could mask a bug.

Reply:
Well that behavior is present [here](https://github.com/arduino/FirmwareUploader/pull/61/files#diff-9bcc47c1909291c8177f70e16def7a94afe0d6c3fb6c937a8f2ccc91432b8aecR166).
At the end of the specified numbers of retries the program exits

Reply:
Yes, but that's a different situation from an unhandled error.

Anyway, I am happy with it as is. I think the retries are a very nice improvement. I always wish AVRDUDE had this option.